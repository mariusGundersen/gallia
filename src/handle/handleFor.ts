import { insertAfter } from "../domUtils.js";
import {
  makeObservable,
  ObservableObject,
  ObservationScope,
} from "../observable.js";
import { makeExpressionEvaluator } from "../utils.js";
import { HandleGenerator, Handler } from "./index.js";

interface Source {
  key: unknown;
  index: number;
  value: unknown;
  startOfItem: Node;
  endOfItem: Node;
  destroy: () => void;
}

export function* handleFor(
  element: HTMLTemplateElement,
  walk: Handler
): HandleGenerator {
  const forExpression = element.getAttribute("x-for");
  if (!forExpression) return;

  const [isFor, name, collection] =
    /^(\w+)\s+of\s+(.*)$/.exec(forExpression) ?? [];
  if (!isFor) return;

  const expression = makeExpressionEvaluator(collection);
  const keyExpression = getKeyExpression(element);

  const documentFragment = element.content;
  documentFragment.prepend(
    documentFragment.ownerDocument.createComment(`start of item`)
  );
  documentFragment.append(
    documentFragment.ownerDocument.createComment(`end of item`)
  );

  const handlers = [...walk(documentFragment, 0)];

  yield (node: Node, data: ObservableObject, scope: ObservationScope) => {
    const element = node as HTMLTemplateElement;
    const parent = element.parentNode;
    const before: Node = element.ownerDocument.createComment(
      `for (${forExpression}) {`
    );
    const after: Node = element.ownerDocument.createComment("}");
    parent!.insertBefore(before, element);
    parent!.replaceChild(after, element);

    const oldItems = new Map<unknown, Source>();
    scope.observeAndReact(
      () => Array.from(expression(data)),
      (items) => {
        const oldList = [...oldItems.values()].sort(
          (a, b) => a.index - b.index
        );
        let endOfPreviousItem = before;
        for (let index = 0; index < items.length; index++) {
          const value = items[index];
          const key = keyExpression({ [name]: value, $index: index });
          const oldItem = oldItems.get(key);
          if (!oldItem) {
            // this is a new item
            const clone = documentFragment.cloneNode(true);
            const startOfItem = clone.firstChild as Node;
            startOfItem.textContent = `start of ${index} with key ${key}`;
            const endOfItem = clone.lastChild as Node;
            endOfItem.textContent = `end of ${index}`;

            const source = (makeObservable({
              key,
              value,
              index,
              startOfItem,
              endOfItem,
            }) as unknown) as Source;

            const subData = {
              get [name]() {
                return source.value;
              },
              get $index() {
                return source.index;
              },
              $parent: data,
            };
            const { scope: subScope, destroy } = scope.createSubScope();
            source.destroy = destroy;

            // remember the childNodes so that we can send them to the handlers
            const fragmentAsNode: Node = {
              // we pretend this is a full node, but it isn't.
              // Thats fine, as it will be passed to a function that only
              // deals with childNodes
              childNodes: Array.from(clone.childNodes),
            } as any;

            insertAfter(endOfPreviousItem, clone);

            handlers.forEach((handler) =>
              handler(fragmentAsNode, subData, subScope)
            );

            endOfPreviousItem = endOfItem;
            oldItems.set(key, source);
          } else if (oldItem.index !== index) {
            // item has moved
            oldItem.index = index;
            if (oldItem.value !== value) {
              oldItem.value = value;
            }

            moveItemAfter(
              oldItem.startOfItem,
              oldItem.endOfItem,
              endOfPreviousItem
            );

            oldList.splice(oldList.indexOf(oldItem), 1);
            endOfPreviousItem = oldItem.endOfItem;
          } else {
            // item has not moved
            oldList.splice(oldList.indexOf(oldItem), 1);
            endOfPreviousItem = oldItem.endOfItem;
            if (oldItem.value !== value) {
              oldItem.value = value;
            }
          }
        }

        // remove any remaining items from before
        for (const oldItem of oldList) {
          oldItems.delete(oldItem.key);
          oldItem.destroy();
          removeItem(oldItem.startOfItem, oldItem.endOfItem);
        }
      }
    );
  };
}

const indexKeyExpression = ({ $index }: { $index: number }) => $index;

function getKeyExpression(elm: HTMLTemplateElement) {
  const key = elm.getAttribute("x-key");

  return key ? makeExpressionEvaluator(key) : indexKeyExpression;
}

function moveItemAfter(
  startOfItem: Node,
  endOfItem: Node,
  endOfPreviousItem: Node
) {
  let item = startOfItem;
  while (true) {
    const nextItem = item.nextSibling;
    insertAfter(endOfPreviousItem, item);
    endOfPreviousItem = item;
    if (item === endOfItem) break;
    if (!nextItem) break;
    item = nextItem;
  }
}

function removeItem(startOfItem: Node, endOfItem: Node) {
  let item = startOfItem;
  while (true) {
    const nextItem = item.nextSibling;
    item.parentNode!.removeChild(item);
    if (item === endOfItem) break;
    if (!nextItem) break;
    item = nextItem;
  }
}
