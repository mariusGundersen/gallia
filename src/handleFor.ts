import { insertAfter } from "./domUtils";
import {
  makeObservable,
  ObservableObject,
  ObservationScope,
} from "./observable";
import { makeExpressionEvaluator } from "./utils";

interface Source {
  key: unknown;
  index: number;
  value: unknown;
  startOfItem: Node;
  endOfItem: Node;
  destroy: () => void;
}

export function handleFor(
  element: HTMLTemplateElement,
  data: ObservableObject,
  scope: ObservationScope,
  walk?: (node: Node, data: ObservableObject, scope: ObservationScope) => void
) {
  const forExpression = element.getAttribute("x-for");
  if (!forExpression) return;

  const [isFor, name, collection] =
    /^(\w+)\s+of\s+(.*)$/.exec(forExpression) ?? [];
  if (!isFor) return;

  const expression = makeExpressionEvaluator(collection);
  const parent = element.parentNode;
  const before: Node = element.ownerDocument.createComment(
    `for (${forExpression}) {`
  );
  const after: Node = element.ownerDocument.createComment("}");
  parent!.insertBefore(before, element);
  parent!.replaceChild(after, element);

  const keyExpression = getKeyExpression(element);

  const oldItems = new Map<unknown, Source>();
  scope.observeAndReact(
    () => expression(data),
    (items) => {
      const oldList = [...oldItems.values()].sort((a, b) => a.index - b.index);
      let endOfPreviousItem = before;
      let index = 0;
      for (const value of items) {
        const key = keyExpression({ [name]: value, $index: index });
        const oldItem = oldItems.get(key);
        if (oldItem?.index === index) {
          console.log("item has not moved", index, value);
          // item has not moved
          oldList.splice(oldList.indexOf(oldItem), 1);
          endOfPreviousItem = oldItem.endOfItem;
          if (oldItem.value !== value) {
            oldItem.value = value;
          }
        } else if (!oldItem) {
          console.log("new item", index, value);
          // this is a new item
          // insert it at the current position
          const clone = element.content.cloneNode(true);
          const startOfItem = element.ownerDocument.createComment(
            `start of ${index} with key ${key}`
          );
          const endOfItem = element.ownerDocument.createComment(
            `end of ${index}`
          );

          const source = ({
            key,
            value,
            index,
            startOfItem,
            endOfItem,
          } as unknown) as Source;
          const subData = Object.create(
            data,
            makeObservable({
              [name]: {
                get() {
                  return source.value;
                },
              },
              $index: {
                get() {
                  return source.index;
                },
              },
            })
          );
          const { scope: subScope, destroy } = scope.createSubScope();
          source.destroy = destroy;
          walk?.(clone, subData, subScope);
          clone.appendChild(endOfItem);
          clone.insertBefore(startOfItem, clone.firstChild);
          insertAfter(endOfPreviousItem, clone);
          endOfPreviousItem = endOfItem;
          oldItems.set(key, source);
        } else {
          console.log("item has moved from", oldItem.index, "to", index, value);
          // item has moved
          // move dom elements
          oldItem.index = index;
          if (oldItem.value !== value) {
            oldItem.value = value;
          }

          let item = oldItem.startOfItem;
          do {
            const nextItem = item.nextSibling;
            insertAfter(endOfPreviousItem, item);
            endOfPreviousItem = item;
            if (!nextItem) break;
            item = nextItem;
          } while (item !== oldItem.endOfItem);

          oldList.splice(oldList.indexOf(oldItem), 1);
        }

        index++;
      }

      // remove any remaining items from before
      for (const oldItem of oldList) {
        console.log("item has been removed", oldItem.index);
        oldItems.delete(oldItem.key);
        oldItem.destroy();
        let item = oldItem.startOfItem;
        do {
          const nextItem = item.nextSibling;
          item.parentNode!.removeChild(item);
          if (!nextItem) break;
          item = nextItem;
        } while (item !== oldItem.endOfItem);
      }
    }
  );
}

const indexKeyExpression = ({ $index }: { $index: number }) => $index;

function getKeyExpression(elm: HTMLTemplateElement) {
  const key = elm.getAttribute("x-key");

  return key ? makeExpressionEvaluator(key) : indexKeyExpression;
}
