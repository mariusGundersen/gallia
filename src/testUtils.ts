export function createElementFromHTML(htmlString: string) {
  var div = document.createElement("div");
  div.innerHTML = htmlString.trim();
  return div.firstChild as HTMLElement;
}
