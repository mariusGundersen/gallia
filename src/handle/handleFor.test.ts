import {
  globalObservationScope,
  makeObservable,
  ObservableObject,
  ObservationScope,
} from "../observable";
import { createElementFromHTML } from "../testUtils";
import { handleFor } from "./handleFor";

describe("for", () => {
  test("empty list", () => {
    const parent = createElementFromHTML(`
      <div>
        <template x-for="item of list">
          Something
        </template>
      </div>
    `);

    const data = {
      list: [],
    };

    const spy = jest.fn();
    const handlerSpy = jest.fn(function () {
      return spy;
    });

    const [result] = [
      ...handleFor(parent.firstElementChild as HTMLTemplateElement, handlerSpy),
    ];

    result(parent.firstElementChild as Node, data, globalObservationScope);

    expect(parent.innerHTML).toMatchSnapshot();

    expect(spy).not.toHaveBeenCalled();
  });

  test("one item", () => {
    const parent = createElementFromHTML(`
      <div>
        <template x-for="item of list">
          Something
        </template>
      </div>
    `);

    const data = {
      list: [1],
    };

    const spy = jest.fn();
    const handlerSpy = jest.fn(function () {
      return (node: Node, data: ObservableObject, scope: ObservationScope) =>
        spy(node.childNodes[0], data, scope);
    });

    const [result] = [
      ...handleFor(parent.firstElementChild as HTMLTemplateElement, handlerSpy),
    ];

    result(parent.firstElementChild as Node, data, globalObservationScope);

    expect(parent.innerHTML).toMatchSnapshot();

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy.mock.calls[0][0]).toBe(
      parent.firstChild?.nextSibling?.nextSibling
    );
    expect(spy.mock.calls[0][1].list).toBe(data.list);
    expect(spy.mock.calls[0][1].$parent.list).toBe(data.list);
    expect(spy.mock.calls[0][1].item).toBe(1);
    expect(spy.mock.calls[0][1].$index).toBe(0);
  });

  test("five items", () => {
    const parent = createElementFromHTML(`
      <div>
        <template x-for="item of list">
          Something
        </template>
      </div>
    `);

    const data = {
      list: [1, 2, 3, 4, 5],
    };

    const spy = jest.fn();
    const handlerSpy = jest.fn(function () {
      return spy;
    });

    const [result] = [
      ...handleFor(parent.firstElementChild as HTMLTemplateElement, handlerSpy),
    ];

    result(parent.firstElementChild as Node, data, globalObservationScope);

    expect(parent.innerHTML).toMatchSnapshot();

    expect(spy).toHaveBeenCalledTimes(5);
  });

  test("iterator with five items", () => {
    const parent = createElementFromHTML(`
      <div>
        <template x-for="item of list()">
          Something
        </template>
      </div>
    `);

    const data = {
      *list() {
        for (let i = 0; i < 5; i++) {
          yield i;
        }
      },
    };

    const [result] = [
      ...handleFor(
        parent.firstElementChild as HTMLTemplateElement,
        createEmptyWalker
      ),
    ];

    result(parent.firstElementChild as Node, data, globalObservationScope);
    expect(parent.innerHTML).toMatchSnapshot();
  });

  test("replacing items", () => {
    const parent = createElementFromHTML(`
      <div>
        <template x-for="item of list">
          Something
        </template>
      </div>
    `);

    const data = makeObservable({
      list: [1],
    });

    const [result] = [
      ...handleFor(
        parent.firstElementChild as HTMLTemplateElement,
        createEmptyWalker
      ),
    ];

    result(parent.firstElementChild as Node, data, globalObservationScope);

    expect(parent.innerHTML).toMatchSnapshot();

    data.list = [1, 2];

    expect(parent.innerHTML).toMatchSnapshot();
  });

  test("appending items", () => {
    const parent = createElementFromHTML(`
      <div>
        <template x-for="item of list">
          Something
        </template>
      </div>
    `);

    const data = {
      list: makeObservable([1]),
    };

    const [result] = [
      ...handleFor(
        parent.firstElementChild as HTMLTemplateElement,
        createEmptyWalker
      ),
    ];

    result(parent.firstElementChild as Node, data, globalObservationScope);

    expect(parent.innerHTML).toMatchSnapshot();

    data.list.push(2);

    expect(parent.innerHTML).toMatchSnapshot();
  });

  test("move item", () => {
    const parent = createElementFromHTML(`
      <div>
        <template x-for="item of list" x-key="item.key">
          Something
        </template>
      </div>
    `);

    const data = makeObservable({
      list: [{ key: 1 }, { key: 2 }],
    });

    const [result] = [
      ...handleFor(
        parent.firstElementChild as HTMLTemplateElement,
        createEmptyWalker
      ),
    ];

    result(parent.firstElementChild as Node, data, globalObservationScope);

    expect(parent.innerHTML).toMatchSnapshot();

    data.list = [{ key: 2 }, { key: 1 }];

    expect(parent.innerHTML).toMatchSnapshot();
  });

  test("remove item at end", () => {
    const parent = createElementFromHTML(`
      <div>
        <template x-for="item of list" x-key="item.key">
          Something
        </template>
      </div>
    `);

    const data = makeObservable({
      list: [{ key: 1 }, { key: 2 }, { key: 3 }],
    });

    const [result] = [
      ...handleFor(
        parent.firstElementChild as HTMLTemplateElement,
        createEmptyWalker
      ),
    ];

    result(parent.firstElementChild as Node, data, globalObservationScope);

    expect(parent.innerHTML).toMatchSnapshot();

    data.list = [{ key: 1 }, { key: 2 }];

    expect(parent.innerHTML).toMatchSnapshot();
  });

  test("remove item at beginning", () => {
    const parent = createElementFromHTML(`
      <div>
        <template x-for="item of list" x-key="item.key">
          Something
        </template>
      </div>
    `);

    const data = makeObservable({
      list: [{ key: 1 }, { key: 2 }, { key: 3 }],
    });

    const [result] = [
      ...handleFor(
        parent.firstElementChild as HTMLTemplateElement,
        createEmptyWalker
      ),
    ];

    result(parent.firstElementChild as Node, data, globalObservationScope);

    expect(parent.innerHTML).toMatchSnapshot();

    data.list = [{ key: 2 }, { key: 3 }];

    expect(parent.innerHTML).toMatchSnapshot();
  });
});

function createEmptyWalker(node: Node) {
  return () => {};
}
