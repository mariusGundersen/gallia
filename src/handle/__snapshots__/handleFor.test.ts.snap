// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`for appending items 1`] = `
"
        <!--for (item of list) {--><!--start of 0 with key 0-->
          Something
        <!--end of 0--><!--}-->
      "
`;

exports[`for appending items 2`] = `
"
        <!--for (item of list) {--><!--start of 0 with key 0-->
          Something
        <!--end of 0--><!--start of 1 with key 1-->
          Something
        <!--end of 1--><!--}-->
      "
`;

exports[`for empty list 1`] = `
"
        <!--for (item of list) {--><!--}-->
      "
`;

exports[`for five items 1`] = `
"
        <!--for (item of list) {--><!--start of 0 with key 0-->
          Something
        <!--end of 0--><!--start of 1 with key 1-->
          Something
        <!--end of 1--><!--start of 2 with key 2-->
          Something
        <!--end of 2--><!--start of 3 with key 3-->
          Something
        <!--end of 3--><!--start of 4 with key 4-->
          Something
        <!--end of 4--><!--}-->
      "
`;

exports[`for iterator with five items 1`] = `
"
        <!--for (item of list()) {--><!--start of 0 with key 0-->
          Something
        <!--end of 0--><!--start of 1 with key 1-->
          Something
        <!--end of 1--><!--start of 2 with key 2-->
          Something
        <!--end of 2--><!--start of 3 with key 3-->
          Something
        <!--end of 3--><!--start of 4 with key 4-->
          Something
        <!--end of 4--><!--}-->
      "
`;

exports[`for move item 1`] = `
"
        <!--for (item of list) {--><!--start of 0 with key 1-->
          Something
        <!--end of 0--><!--start of 1 with key 2-->
          Something
        <!--end of 1--><!--}-->
      "
`;

exports[`for move item 2`] = `
"
        <!--for (item of list) {--><!--start of 1 with key 2-->
          Something
        <!--end of 1--><!--start of 0 with key 1-->
          Something
        <!--end of 0--><!--}-->
      "
`;

exports[`for one item 1`] = `
"
        <!--for (item of list) {--><!--start of 0 with key 0-->
          Something
        <!--end of 0--><!--}-->
      "
`;

exports[`for remove item at end 1`] = `
"
        <!--for (item of list) {--><!--start of 0 with key 1-->
          Something
        <!--end of 0--><!--start of 1 with key 2-->
          Something
        <!--end of 1--><!--start of 2 with key 3-->
          Something
        <!--end of 2--><!--}-->
      "
`;

exports[`for remove item at end 2`] = `
"
        <!--for (item of list) {--><!--start of 0 with key 1-->
          Something
        <!--end of 0--><!--start of 1 with key 2-->
          Something
        <!--end of 1--><!--}-->
      "
`;

exports[`for replacing items 1`] = `
"
        <!--for (item of list) {--><!--start of 0 with key 0-->
          Something
        <!--end of 0--><!--}-->
      "
`;

exports[`for replacing items 2`] = `
"
        <!--for (item of list) {--><!--start of 0 with key 0-->
          Something
        <!--end of 0--><!--start of 1 with key 1-->
          Something
        <!--end of 1--><!--}-->
      "
`;
